search/pre_ranker.cpp

```diff
@@ -171,28 +179,36 @@ void PreRanker::Filter(bool viewportSearch)
 
   base::SortUnique(m_results, lessForUnique, base::EqualsBy(&PreRankerResult::GetId));
 
-  if (viewportSearch)
+  if (m_params.m_viewportSearch)
     FilterForViewportSearch();
 
+  // Viewport search ends here.
   if (m_results.size() <= BatchSize())
     return;
 
-  vector<size_t> indices(m_results.size());
-  generate(indices.begin(), indices.end(), [n = 0] () mutable { return n++; });
-  unordered_set<size_t> filtered;
+  std::vector<size_t> indices(m_results.size());
+  std::generate(indices.begin(), indices.end(), [n = 0] () mutable { return n++; });
+  std::unordered_set<size_t> filtered;
 
   auto const iBeg = indices.begin();
   auto const iMiddle = iBeg + BatchSize();
   auto const iEnd = indices.end();
 
-  nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessDistance, m_results));
+  std::nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessDistance, m_results));
   filtered.insert(iBeg, iMiddle);
 
   if (!m_params.m_categorialRequest)
   {
-    nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessRankAndPopularity, m_results));
+    std::nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessRankAndPopularity, m_results));
     filtered.insert(iBeg, iMiddle);
-    nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessByExactMatch, m_results));
+
+    // Shuffle to give a chance to far results, not only closest ones (see above).
+    // Search is not stable in rare cases, but we avoid increasing m_everywhereBatchSize.
+    /// @todo Move up, when we will have _enough_ ranks and popularities.
+    std::random_device rd;
```

Comment:
Move the device and engine to a variable that is initialized once. It may be _very_ slow on some platforms to recreate them.
https://www.pcg-random.org/posts/cpps-random_device.html